presto-main/src/main/java/com/facebook/presto/server/TaskInfoResource.java

```diff
@@ -13,39 +13,88 @@
 
 import javax.annotation.security.RolesAllowed;
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.DefaultValue;
 import javax.ws.rs.GET;
+import javax.ws.rs.HeaderParam;
 import javax.ws.rs.NotFoundException;
 import javax.ws.rs.Path;
 import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.container.AsyncResponse;
+import javax.ws.rs.container.Suspended;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.UriInfo;
 
+import java.net.URI;
 import java.util.Optional;
 
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.net.HttpHeaders.X_FORWARDED_PROTO;
 import static java.util.Objects.requireNonNull;
+import static javax.ws.rs.core.Response.Status.SERVICE_UNAVAILABLE;
 
 @Path("v1/taskInfo")
 @RolesAllowed("ADMIN")
 public class TaskInfoResource
 {
+    private final DispatchManager dispatchManager;
     private final QueryManager queryManager;
+    private final boolean resourceManagerEnabled;
+    private final InternalNodeManager internalNodeManager;
+    private final Optional<ResourceManagerProxy> proxyHelper;
 
     @Inject
-    public TaskInfoResource(QueryManager queryManager)
+    public TaskInfoResource(
+            DispatchManager dispatchManager,
+            QueryManager queryManager,
+            InternalNodeManager internalNodeManager,
+            ServerConfig serverConfig,
+            Optional<ResourceManagerProxy> proxyHelper)
     {
+        this.dispatchManager = requireNonNull(dispatchManager, "dispatchManager is null");
         this.queryManager = requireNonNull(queryManager, "queryManager is null");
+        this.internalNodeManager = requireNonNull(internalNodeManager, "internalNodeManager is null");
+        this.resourceManagerEnabled = requireNonNull(serverConfig, "serverConfig is null").isResourceManagerEnabled();
+        this.proxyHelper = requireNonNull(proxyHelper, "proxyHelper is null");
     }
 
     @GET
     @Path("{taskId}")
-    public TaskInfo getTaskInfo(@PathParam("taskId") TaskId taskId)
-            throws NotFoundException
+    @Produces(MediaType.APPLICATION_JSON)
+    public void getTaskInfo(@PathParam("taskId") TaskId taskId,
+            @QueryParam("includeLocalQueryOnly") @DefaultValue("false") boolean includeLocalQueryOnly,
+            @HeaderParam(X_FORWARDED_PROTO) String xForwardedProto,
+            @Context UriInfo uriInfo,
+            @Context HttpServletRequest servletRequest,
+            @Suspended AsyncResponse asyncResponse)
+            throws WebApplicationException
+    {
+        if (requestNeedsToBeProxied(taskId, includeLocalQueryOnly)) {
+            proxyTaskInfo(servletRequest, asyncResponse, xForwardedProto, uriInfo);
+        }
+        else {
+            asyncResponse.resume(Response.ok(getTaskInfo(taskId)).build());
```

Comment:
Response.ok is not correct here since getTaskInfo can throw a NotFoundException.
you need to respond with   `asyncResponse.resume(Response.status(NOT_FOUND).build());` for that case.